### PR TITLE
KafkaTopic/KafkaUser CRD Support

### PR DESCRIPTION
**Pull request by tinyzimmer**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | maybe? `cert-manager` dependency in current state
| Deprecations?   | no
| Related tickets | #46 #47 
| License         | Apache 2.0

### What's in this PR?

This PR provides `KafkaUser` and `KafkaTopic` CRDs. Additionally, it allows the user the option to have the controller create a full PKI for the cluster backed by `cert-manager`. 

If the user still provides their own secret, a `ClusterIssuer` is made from that secret for provisioning kafka users. This introduces a hard-dependency on `cert-manager` for SSL support if it wasn't already there. I'm contemplating adding the logic to have the controller carry out all the CA functions. It would not be that difficult. 

Below are examples of the CRs:

```yaml
apiVersion: banzaicloud.banzaicloud.io/v1alpha1
kind: KafkaTopic
metadata:
  name: test-topic
  namespace: kafka
spec:
  clusterRef:
    name: kafka
  name: test-topic
  partitions: 6
  replicationFactor: 2
  config:
    "retention.ms": "800000"
    "cleanup.policy": "delete"
---
apiVersion: banzaicloud.banzaicloud.io/v1alpha1
kind: KafkaUser
metadata:
  name: test-kafkauser
  namespace: kafka
spec:
  clusterRef:
    name: kafka
  name: test-kafkauser
  secretName: test-kafkauser-secret
  includeJKS: true
  topicGrants:
    - topicName: test-topic
      accessType: read
    - topicName: test-topic
      accessType: write

```

### Why?

Because awesomeness

### Additional context

I've tested various SSL KafkaCluster configurations and all appears to work as expected. I am able to then mount my `KafkaUser` secrets inside a pod and produce/consume from Kafka using both the console producer and consumer. 

I was having some issue with testing that could have been related to my additions, or possibly just the environment I was testing against. `Namespace` kept being an empty string when trying to create clusters. Then the reconciling steps for envoy kept failing due to "namespace being required." Removing `//+kubebuilder:resource:path=kafkaclusters,scope=Cluster` from the type fixed this for me, but it's possible there is a better fix. 

### Checklist

- [X] Implementation tested
- [X] Error handling code meets the [guideline] - _maybe? you may see some places for improvement_
- [X] Logging code meets the guideline - _same as error handling_

### To Do

- [x] User guide and development docs updated 
- [x] Update helm chart with webhook-service, potential new dependencies, and new CRDs
- [x] There is some leftover `operator-sdk` commentary that I should prune out. I haven't used `kubebuilder` before so still learning.
- [x] There is some redundant code about now, mainly due to the new `kafkautil` package addition.
- [x] I should probably give the cluster ownership over users/topics made in reference to it. That way they can be properly garbage collected on deleting a `kafkacluster`. 
- [x] Increase memory limits in `config/manager/manager.yaml`? They are realllly low.

### CODE REVIEW

**Comment by tinyzimmer** [comment]
I see the me not copying licensing info made ci mad. I can take care of that later.

File: controllers/kafkatopic_controller.go [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// we got a topic back
+	if existing != nil {
+		// check if requesting a partition decrease, we can't do this
+		reqLogger.Info("Topic already exists, verifying configuration")
+		if existing.NumPartitions > instance.Spec.Partitions {
+			reqLogger.Info(fmt.Sprint("IGNORING: Spec is requesting partition decrease from", existing.NumPartitions, "to", instance.Spec.Partitions))
```
Comment (tinyzimmer) [inline controllers/kafkatopic_controller.go]: I think I need to make sure I don't requeue here

**Comment by tinyzimmer** [comment]
Another thought, since I am canning the ACLs. I am almost wondering if it makes more sense to have distinct `KafkaConsumer` and `KafkaProducer` CRs. Then instead of `topicGrants` they would each just have a list of `topics`. 

A later implementation could have a `KafkaAdmin` that could be used for all the various off-the-shelf manager utilities that teams use for development.

**Comment by tinyzimmer** [comment]
One more thought before I take off for the night - Should I make the `KafkaTopic` names become the actual topic name and then remove the `name` field from the spec?

Considerations

If I don't, I need to implement logic to make sure we don't write over an existing topic with a new CR using the same name.

If I do, the user will not be able to use certain characters in their topic names. For example, underscores (_).

**Comment by baluchicken** [comment]
Hi @tinyzimmer Thanks for this awesome PR, I just ran through and looks cool. I will have time to review it, during the second half of the week. 

About the empty namespace issue, I mistakenly added `//+kubebuilder:resource:path=kafkaclusters,scope=Cluster` but it is not required.

**Comment by tinyzimmer** [comment]
I've done most of the cleanup I had originally planned for. I'm beating my head on one more issue that I can't decide the best way to approach.

1. User creates `KafkaTopic`
2. User deletes `KafkaCluster` - _causes previous KafkaTopic CR to be deleted_
3. User creates new `KafkaCluster` pointing at same zookeeper as last one
4. User attempts to recreate `KafkaTopic` and webhook denies them because zookeeper already has the topic in its store.

**Comment by baluchicken** [comment]
> I've done most of the cleanup I had originally planned for. I'm beating my head on one more issue that I can't decide the best way to approach.
> 
> 1. User creates `KafkaTopic`
> 2. User deletes `KafkaCluster` - _causes previous KafkaTopic CR to be deleted_
> 3. User creates new `KafkaCluster` pointing at same zookeeper as last one
> 4. User attempts to recreate `KafkaTopic` and webhook denies them because zookeeper already has the topic in its store.
> 
> I'm almost wondering if I want to leverage the zookeeper associated with the kafka cluster to store some operator runtime data and have it "send messages" to itself. For example, place a key that references topics to a cluster, then on the admission-webhook, use that to see if the `KafkaCluster` for that topic still exists. If it's gone "release the lock". Something about this idea feels very ugly though.

I don't think we should interact with the zookeeper at all. 
If I understand the code well, when the cluster is removed, the topic controller simply removes the finaliser from the KafkaTopics owned by the cluster so because of the owner reference Kubernetes will handle the rest and delete the CRs.
I think what we should do is the following, we should place a finaliser on the KafkaCluster CR so Kubernetes will not delete it immediately. Before the Operator removes the KafkaCluster CR-s finaliser it should delete all associated KafkaTopics,KafkaUser CRs. That way the Topics/Users will be deleted before the cluster. I think this is our only chance since, we need the KafkaCluster up and running to be able to delete topics. Kubernetes owner reference works in a way that it only triggers the "childrens" delete when the owner is deleted, but it is too late for us.

**Comment by tinyzimmer** [comment]
Yea that makes sense. I'll take a look at it tonight and see what I can do.

File: pkg/resources/pki/certs.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// A full PKI for Kafka
+
+func (r *Reconciler) kafkapki() ([]runtime.Object, error) {
+	rootCertMeta := templates.ObjectMeta(fmt.Sprintf(brokerCACertTemplate, r.KafkaCluster.Name), labelsForKafkaPKI(r.KafkaCluster.Name), r.KafkaCluster)
+	rootCertMeta.Namespace = "cert-manager"
+
+	if r.KafkaCluster.Spec.ListenersConfig.SSLSecrets.Create {
```
Comment (pbalogh-sa) [inline pkg/resources/pki/certs.go]: If the `sslSecrets` isn't defined in CR because you want to try without ssl, this will be a nil pointer dereference.

Comment (tinyzimmer) [reply pkg/resources/pki/certs.go]: Good catch, I think I might have that problem in the end of `pki.go` also. Will fix.

Comment (tinyzimmer) [reply pkg/resources/pki/certs.go]: I fixed this by just returning at the beginning of the Reconcile if `sslSecrets` is nil. I also found an issue with when I took out a bunch of hardcoded strings and made them constants. I left one wrong by mistake.

**Pull request merged**